File: playlist_builder/ui/shared/autocomplete/keyword_suggestion.py
from __future__ import annotations

import re
import unicodedata

_STOP_WORDS = {
    "a",
    "an",
    "and",
    "de",
    "du",
    "feat",
    "featuring",
    "in",
    "la",
    "le",
    "les",
    "of",
    "the",
    "with",
    "ft",
    "vs",
}

_MOOD_KEYWORDS = {
    "calm": ("calm", "chill", "relax", "soft"),
    "dramatic": ("dramatic", "epic", "orchestral"),
    "energetic": ("energy", "energetic", "intense", "power"),
    "happy": ("happy", "joy", "sunshine", "smile"),
    "melancholic": ("sad", "lonely", "tears", "blue"),
    "romantic": ("love", "heart", "kiss", "romance"),
    "summer": ("summer", "sun", "beach", "festival"),
    "upbeat": ("upbeat", "party", "dance", "groove"),
}

_GENRE_HINTS = {
    "alternative": ("alternative", "alt rock", "indie"),
    "dance": ("dance", "edm", "house", "club"),
    "electronic": ("electronic", "electro", "synth"),
    "hip-hop": ("hip hop", "rap", "trap"),
    "pop": ("pop",),
    "rock": ("rock", "metal", "punk"),
}


def suggest_keywords(
    *,
    artist_name: str = "",
    artist_type: str = "",
    track_title: str = "",
    album_title: str = "",
    release_year: int | None = None,
    primary_genre_name: str = "",
) -> list[str]:
    """Suggest playlist keywords from autocomplete metadata without provider coupling."""
    suggestions: list[str] = []

    def add(value: str) -> None:
        normalized = _normalize_keyword(value)
        if not normalized or normalized in suggestions:
            return
        suggestions.append(normalized)

    if primary_genre_name.strip():
        add(primary_genre_name)

    for keyword in _genre_hints_from_text(primary_genre_name):
        add(keyword)

    for keyword in _title_keywords(track_title):
        add(keyword)

    if album_title.strip():
        for keyword in _title_keywords(album_title, max_tokens=2):
            add(keyword)

    if release_year is not None and release_year >= 1900:
        decade = (release_year // 10) * 10
        add(f"{decade}s")

    for mood, triggers in _MOOD_KEYWORDS.items():
        haystack = f"{track_title} {album_title}".casefold()
        if any(trigger in haystack for trigger in triggers):
            add(mood)

    if artist_type.strip().casefold() in {"band", "group", "orchestra"}:
        add(artist_type.strip().casefold())

    # Artist name is a weak signal; only use it when nothing else was inferred.
    if not suggestions and artist_name.strip():
        for keyword in _title_keywords(artist_name, max_tokens=1):
            add(keyword)

    return suggestions[:8]


def _genre_hints_from_text(value: str) -> list[str]:
    lowered = value.casefold()
    hints: list[str] = []
    for keyword, triggers in _GENRE_HINTS.items():
        if any(trigger in lowered for trigger in triggers):
            hints.append(keyword)
    return hints


def _title_keywords(value: str, *, max_tokens: int = 3) -> list[str]:
    cleaned = re.sub(r"\([^)]*\)", " ", value)
    cleaned = re.sub(r"\[[^]]*\]", " ", cleaned)
    cleaned = re.sub(r"(?i)\bfeat\.?.*$", " ", cleaned)
    tokens = [_normalize_keyword(token) for token in re.split(r"[^a-zA-Z0-9]+", cleaned)]
    result: list[str] = []
    for token in tokens:
        if not token or token in _STOP_WORDS or len(token) < 3:
            continue
        if token not in result:
            result.append(token)
        if len(result) >= max_tokens:
            break
    return result


def _normalize_keyword(value: str) -> str:
    normalized = unicodedata.normalize("NFKD", value.strip().casefold())
    normalized = "".join(char for char in normalized if not unicodedata.combining(char))
    normalized = re.sub(r"[^a-z0-9]+", " ", normalized).strip()
    return normalized

File: apps/resonance/ResonanceCore/Sources/ResonanceCore/SmartInput/KeywordSuggestion.swift
import Foundation

public enum KeywordSuggestionSource: String, Sendable {
    case manual
    case automatic
}

public struct KeywordSuggestionInput: Sendable {
    public var artistName: String
    public var artistType: String
    public var trackTitle: String
    public var albumTitle: String
    public var releaseYear: Int?
    public var primaryGenreName: String

    public init(
        artistName: String = "",
        artistType: String = "",
        trackTitle: String = "",
        albumTitle: String = "",
        releaseYear: Int? = nil,
        primaryGenreName: String = ""
    ) {
        self.artistName = artistName
        self.artistType = artistType
        self.trackTitle = trackTitle
        self.albumTitle = albumTitle
        self.releaseYear = releaseYear
        self.primaryGenreName = primaryGenreName
    }

    public init(artist: ArtistRef?, track: TrackRef?) {
        artistName = artist?.displayName ?? track?.artistName ?? ""
        artistType = artist?.artistType ?? ""
        trackTitle = track?.title ?? ""
        albumTitle = track?.albumTitle ?? ""
        releaseYear = track?.releaseYear
        primaryGenreName = track?.genreName ?? ""
    }
}

public enum KeywordSuggestionEngine {
    private static let stopWords: Set<String> = [
        "a", "an", "and", "de", "du", "feat", "featuring", "in", "la", "le", "les", "of", "the", "with", "ft", "vs",
    ]

    private static let moodKeywords: [String: [String]] = [
        "calm": ["calm", "chill", "relax", "soft"],
        "dramatic": ["dramatic", "epic", "orchestral"],
        "energetic": ["energy", "energetic", "intense", "power"],
        "happy": ["happy", "joy", "sunshine", "smile"],
        "melancholic": ["sad", "lonely", "tears", "blue"],
        "romantic": ["love", "heart", "kiss", "romance"],
        "summer": ["summer", "sun", "beach", "festival"],
        "upbeat": ["upbeat", "party", "dance", "groove"],
    ]

    private static let genreHints: [String: [String]] = [
        "alternative": ["alternative", "alt rock", "indie"],
        "dance": ["dance", "edm", "house", "club"],
        "electronic": ["electronic", "electro", "synth"],
        "hip-hop": ["hip hop", "rap", "trap"],
        "pop": ["pop"],
        "rock": ["rock", "metal", "punk"],
    ]

    public static func suggestKeywords(from input: KeywordSuggestionInput) -> [String] {
        var suggestions: [String] = []

        func add(_ value: String) {
            let normalized = normalizeKeyword(value)
            guard !normalized.isEmpty, !suggestions.contains(normalized) else { return }
            suggestions.append(normalized)
        }

        if !input.primaryGenreName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            add(input.primaryGenreName)
            for keyword in genreHints(from: input.primaryGenreName) {
                add(keyword)
            }
        }

        for keyword in titleKeywords(input.trackTitle) {
            add(keyword)
        }

        if !input.albumTitle.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            for keyword in titleKeywords(input.albumTitle, maxTokens: 2) {
                add(keyword)
            }
        }

        if let releaseYear = input.releaseYear, releaseYear >= 1900 {
            let decade = (releaseYear / 10) * 10
            add("\(decade)s")
        }

        let haystack = "\(input.trackTitle) \(input.albumTitle)".lowercased()
        for (mood, triggers) in moodKeywords {
            if triggers.contains(where: { haystack.contains($0) }) {
                add(mood)
            }
        }

        let artistType = input.artistType.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if ["band", "group", "orchestra"].contains(artistType) {
            add(artistType)
        }

        if suggestions.isEmpty, !input.artistName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            for keyword in titleKeywords(input.artistName, maxTokens: 1) {
                add(keyword)
            }
        }

        return Array(suggestions.prefix(8))
    }

    public static func automaticKeywordRefs(from input: KeywordSuggestionInput) -> [KeywordRef] {
        suggestKeywords(from: input).map { label in
            KeywordRef(
                id: "auto:\(label.lowercased().replacingOccurrences(of: " ", with: "-"))",
                label: label,
                source: .automatic
            )
        }
    }

    public static func mergeAutomaticKeywords(
        existing: [KeywordRef],
        suggested: [KeywordRef]
    ) -> [KeywordRef] {
        var merged = existing
        let manualLabels = Set(existing.filter { $0.source == .manual }.map { normalizeKeyword($0.label) })
        let existingLabels = Set(existing.map { normalizeKeyword($0.label) })

        for keyword in suggested {
            let normalized = normalizeKeyword(keyword.label)
            guard !normalized.isEmpty else { continue }
            if manualLabels.contains(normalized) || existingLabels.contains(normalized) {
                continue
            }
            merged.append(keyword)
        }
        return merged
    }

    private static func genreHints(from value: String) -> [String] {
        let lowered = value.lowercased()
        return genreHints.compactMap { keyword, triggers in
            triggers.contains(where: { lowered.contains($0) }) ? keyword : nil
        }
    }

    private static func titleKeywords(_ value: String, maxTokens: Int = 3) -> [String] {
        var cleaned = value
        cleaned = cleaned.replacingOccurrences(of: #"\([^)]*\)"#, with: " ", options: .regularExpression)
        cleaned = cleaned.replacingOccurrences(of: #"\[[^]]*\]"#, with: " ", options: .regularExpression)
        cleaned = cleaned.replacingOccurrences(of: #"(?i)\bfeat\.?.*$"#, with: " ", options: .regularExpression)

        let tokens = cleaned
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .map(normalizeKeyword)
            .filter { !$0.isEmpty && !stopWords.contains($0) && $0.count >= 3 }

        var result: [String] = []
        for token in tokens where !result.contains(token) {
            result.append(token)
            if result.count >= maxTokens {
                break
            }
        }
        return result
    }

    private static func normalizeKeyword(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
            .replacingOccurrences(of: #"[^a-z0-9]+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
